src/toil_lib/programs.py

```diff
@@ -124,3 +125,282 @@ def _fix_permissions(base_docker_call, tool, work_dir):
     stat = os.stat(work_dir)
     command = base_docker_call + [tool] + ['-R', '{}:{}'.format(stat.st_uid, stat.st_gid), '/data']
     subprocess.check_call(command)
+
+
+class Docker(object):
+    """
+    A builder (https://en.wikipedia.org/wiki/Builder_pattern) for Popen objects representing a
+    Docker client invocation. When building a Popen object, it distinguishes between three
+    layers, the `popen` layer, the `docker` layer and the `extra` layer. The `popen` layer
```

Comment:
I agree `extra` is bad and I like just leaving it off. So 

extra_command => command
extra_params =>params

but there also the `extra(command,*args,**opts)` shortcut. What do we name that?